### PR TITLE
Create basic Cloud Connect GCP configuration

### PR DESCRIPTION
**Pull request by rohrlaf**
- adds basic GCP configuration for Cloud Connect according to https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-gcp
- splits module and example like existing AWS scripts
- updates README file to distinguish between AWS and GCP
- align firewall descriptions for AWS and GCP with Developer Docs tutorials

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,47 +1,71 @@
...
+  - Secret Access Key
+  - Security Group Id
 
 The outputs that are produced by the script can be used in a next step to configure the [Bitmovin Encoder for Cloud Connect](https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-aws#configure-your-bitmovin-account).
 
+### GCP
+
+The GCP module will create the following resources to enable running encoding jobs with Bitmovin Cloud Connect:
+
+- Enable the Compute Engine API
```
Comment (simontischler) [inline README.md]: Nitpick: Either do Enable**s** here or remove the trailing **s** on the other items, to keep it consistent.

File: README.md [hunk 2]
```diff
@@ -50,64 +74,78 @@ provider "aws" {
...
-````
+```sh
 terraform destroy
-````
+```
 
-**Warning**: using the above command will remove the user, policy, security group and rules needed for the Bitmovin Cloud Connect product to work. Use the destroy command when you want to stop using the previously created resources.
+**Warning**: using the command above will remove the created Terraform-managed resources for Bitmovin Cloud Connect to work. Use the `destroy` command only when you are sure you want to stop using the previously created resources.
 
 ## Inputs
```
Comment (simontischler) [inline README.md:120]: Inputs should be on a different header-level than the following entries.

File: examples/gcp/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+output "private_key_instructions" {
+  value = "The see your private key, run `terraform output -json`. Handle credentials carefully."
+}
+
+output "network_id" {
+  value = module.bitmovin_cloud_connect.network_id
+}
+
+output "subnet_id_instructions" {
+  value = "The find your region's subnet id, please go to https://console.cloud.google.com/networking/networks/list"
```
Comment (simontischler) [inline examples/gcp/outputs.tf]: Shouldn't it be To instead of The?

File: examples/gcp/terraform.tf [hunk 4]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.0"
+
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "5.35.0"
```
Comment (simontischler) [inline examples/gcp/terraform.tf]: We often do [version-constraints](https://developer.hashicorp.com/terraform/language/expressions/version-constraints#version-constraint-syntax) using the pessimistic constraint operator, to allow at least patch-versions. Did you think about locking it to the specific version? Just out of curiosity.

File: modules/gcp/rules.tf [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+# create auto-mode VPC
+resource "google_compute_network" "bitmovin_vpc_network" {
+  name = var.network_name
+
+  depends_on = [google_project_service.compute-engine]
+}
+
+resource "google_compute_firewall" "bitmovin_allow_internal" {
+  description   = "For communication between the session manager VM instance and its instance manager VM instances"
```
Comment (simontischler) [inline modules/gcp/rules.tf:20]: Should we maybe align this with AWS, I suspect, this should be pretty much the same as
`Allow communication between the main VM instance and the worker VM instances`

File: modules/gcp/rules.tf [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  allow {
+    protocol = "udp"
+    ports    = ["0-65535"]
+  }
+}
+
+resource "google_compute_firewall" "bitmovin_allow_ssh" {
+  description   = "For incoming commands (i.e. pulling and starting docker containers)"
```
Comment (simontischler) [inline modules/gcp/rules.tf]: Maybe we can expose less information here and do something like
`For incoming commands from the bitmovin API to control the encoding.`

I don't think it is relevant for the customer that we pull docker images here.

File: examples/gcp/terraform.tf [hunk 7]
```diff
@@ -0,0 +1,10 @@
+terraform {
```
Comment (rohrlaf) [inline examples/gcp/terraform.tf:1]: The current scripts use `versions.tf`, but the official docs use `terraform.tf` for versioning: https://developer.hashicorp.com/terraform/tutorials/configuration-language/versions

File: modules/gcp/rules.tf [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+resource "google_project_service" "compute-engine" {
+  project = var.project_id
+  service = "compute.googleapis.com"
+}
+
+# create auto-mode VPC
+resource "google_compute_network" "bitmovin_vpc_network" {
+  name = var.network_name
+
+  depends_on = [google_project_service.compute-engine]
```
Comment (rohrlaf) [inline modules/gcp/rules.tf:16]: Ran into race conditions, as the VPC network requires the enabled Compute Engine API.

File: examples/gcp/main.tf [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+provider "google" {
+  project = local.project_id
+}
+
+module "bitmovin_cloud_connect" {
+  project_id = local.project_id
+  source     = "../../modules/gcp" # local reference when you clone the repository
+  # source     = "github.com/bitmovin/terraform-cloud-connect/blob/main/modules/gcp" # remote GitHub reference
```
Comment (rohrlaf) [inline examples/gcp/main.tf:12]: Unsure, what works best when you clone the repo. But the README refers to the GitHub repo directly.

**Review by simontischler — APPROVED** [review]
Really nice work 💪

File: examples/gcp/terraform.tf (hunk 4 above)
Comment (rohrlaf) [reply examples/gcp/terraform.tf]: @simontischler Guess I just took this from the tutorial and did not consider it further. Will use the pessimistic constraint operator to allow minor versions `~> 5.35` then, when looking at their update history:

image

File: modules/gcp/rules.tf (hunk 5 above)
Comment (rohrlaf) [reply modules/gcp/rules.tf:20]: @simontischler I took the same wording from our tutorials, so I guess AWS should be aligned in the scripts then.

GCP: https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-gcp#configure-and-create-firewall-rules
> For communication between the session manager VM instance and its instance manager VM instances

AWS: https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-aws#create-a-security-group
> For communication between the session manager VM instance and its instance manager VM instances

File: modules/gcp/rules.tf (hunk 6 above)
Comment (rohrlaf) [reply modules/gcp/rules.tf]: @simontischler I agree with you on removing irrelevant information to minimize attack vectors. But this is currently also written in the tutorials as well. My main goal was to stay consistent with our tutorials and allow customers to inspect the scripts against them.

- AWS: https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-aws#create-a-security-group
- GCP: https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-gcp#configure-and-create-firewall-rules

But I can also update all occurrences to your suggestion. 👍

File: README.md (hunk 1 above)
Comment (rohrlaf) [reply README.md]: Thanks, fixed in dbe4e83

File: README.md (hunk 2 above)
Comment (rohrlaf) [reply README.md:120]: Thanks, fixed in dbe4e83

File: examples/gcp/outputs.tf (hunk 3 above)
Comment (rohrlaf) [reply examples/gcp/outputs.tf]: Thanks, fixed in dbe4e83

File: examples/gcp/terraform.tf (hunk 4 above)
Comment (rohrlaf) [reply examples/gcp/terraform.tf]: Done in a2fa0a9

File: modules/gcp/rules.tf (hunk 5 above)
Comment (rohrlaf) [reply modules/gcp/rules.tf:20]: Done in da0ff14

File: modules/gcp/rules.tf (hunk 6 above)
Comment (rohrlaf) [reply modules/gcp/rules.tf]: Done in da0ff14. Updated Developer Docs as well:
- AWS https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-aws#create-a-security-group
- GCP https://developer.bitmovin.com/encoding/docs/using-bitmovin-cloud-connect-with-gcp#configure-and-create-firewall-rules

File: examples/gcp/main.tf (hunk 9 above)
Comment (jamdrop) [reply examples/gcp/main.tf:12]: the reason why we kept the GH link in the example was to make it easier for customers to copy-paste this I don't think they will clone this repo

File: examples/aws/main.tf [hunk 10]
```diff
@@ -3,7 +3,7 @@ provider "aws" {
 }
 
 module "bitmovin_cloud_connect" {
-  source  = "github.com/bitmovin/terraform-cloud-connect//modules/aws"
+  source = "github.com/bitmovin/terraform-cloud-connect/blob/main/modules/aws"
```
Comment (jamdrop) [inline examples/aws/main.tf:6]: why did you opt to change this?

File: modules/gcp/rules.tf [hunk 11]
```diff
@@ -0,0 +1,135 @@
+locals {
+  internal_network_blocks        = ["10.0.0.0/8"]
+  bitmovin_static_network_blocks = ["104.199.97.13/32", "35.205.157.162/32"]
+}
+
+# enable Compute Engine API
+resource "google_project_service" "compute-engine" {
+  project = var.project_id
```
Comment (jamdrop) [inline modules/gcp/rules.tf:8]: the project argument is optional and usually this is taken from the provider that means you could get rid of that variable

File: modules/gcp/rules.tf [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+  name = var.network_name
+
+  depends_on = [google_project_service.compute-engine]
+}
+
+resource "google_compute_firewall" "bitmovin_allow_internal" {
+  description   = "For communication between the session manager VM instance and its instance manager VM instances"
+  name          = "${google_compute_network.bitmovin_vpc_network.name}-allow-internal"
+  network       = google_compute_network.bitmovin_vpc_network.name
+  source_ranges = local.internal_network_blocks
```
Comment (jamdrop) [inline modules/gcp/rules.tf:23]: you could also take this from the `google_compute_network` data source and also get rid of the local for that

**Pull request merged**